### PR TITLE
HAWQ-1228. Use profile based on file format in HCatalog integration(HiveRC, HiveText profiles) for homogeneous tables.

### PR DESCRIPTION
**Pull request by sansanichfb**
Work still in progress, but the main part is done. I want to have earlier feedback on it, it's quite a big diff.

### CODE REVIEW

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java [hunk 1]
```diff
@@ -233,4 +238,9 @@ private void resolvePrimitive(Object o, PrimitiveObjectInspector oi) throws IOEx
         }
         firstColumn = false;
     }
+
+    @Override
+    void parseDelimiterChar(InputData input) {
+        delimiter = 44; //,
```
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java]: Is this a temporary workaround ?
Why not use a static variable to store the delimiter value

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveAccessor.java [hunk 2]
```diff
@@ -138,12 +135,11 @@ protected Object getReader(JobConf jobConf, InputSplit split)
      */
     private InputFormat<?, ?> createInputFormat(InputData input)
             throws Exception {
-        String userData = new String(input.getFragmentUserData());
-        String[] toks = userData.split(HiveDataFragmenter.HIVE_UD_DELIM);
-        initPartitionFields(toks[3]);
-        filterInFragmenter = new Boolean(toks[4]);
+        HiveUserData hiveUserData = HiveUtilities.parseHiveUserData(input);
+        initPartitionFields(hiveUserData.getPartitionKeys());
```
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveAccessor.java]: Will we always have value for partitionKeys ? If not, initParitionFields will throw NPE

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java [hunk 3]
```diff
@@ -57,11 +59,11 @@
  */
 public class HiveColumnarSerdeResolver extends HiveResolver {
     private static final Log LOG = LogFactory.getLog(HiveColumnarSerdeResolver.class);
-    private ColumnarSerDeBase deserializer;
+    //private ColumnarSerDeBase deserializer;
```
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java]: Remove line

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveLineBreakAccessor.java [hunk 4]
```diff
@@ -43,9 +43,9 @@
     public HiveLineBreakAccessor(InputData input) throws Exception {
         super(input, new TextInputFormat());
         ((TextInputFormat) inputFormat).configure(jobConf);
-        String[] toks = HiveInputFormatFragmenter.parseToks(input, PXF_HIVE_SERDES.LAZY_SIMPLE_SERDE.name());
-        initPartitionFields(toks[HiveInputFormatFragmenter.TOK_KEYS]);
-        filterInFragmenter = new Boolean(toks[HiveInputFormatFragmenter.TOK_FILTER_DONE]);
+        HiveUserData hiveUserData = HiveUtilities.parseHiveUserData(input, PXF_HIVE_SERDES.LAZY_SIMPLE_SERDE);
+        initPartitionFields(hiveUserData.getPartitionKeys());
```
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveLineBreakAccessor.java]: Possible NPE in this function

**Comment by shivzone** [comment]
Each property in the profile was meant to be a particular class which allowed any user to customize/extend it. In the same theme, instead of definiging outputFormat in the pxf-profile, lets go with <formatter> and get the classname.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java (hunk 1 above)
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java]: Yes, it's temporary, work is in progress to actually use appropriate parameters from Hive's SerDe.

File: pxf/pxf-service/src/main/resources/pxf-profiles-default.xml [hunk 5]
```diff
@@ -49,6 +49,7 @@ under the License.
             <accessor>org.apache.hawq.pxf.plugins.hive.HiveAccessor</accessor>
             <resolver>org.apache.hawq.pxf.plugins.hive.HiveResolver</resolver>
             <metadata>org.apache.hawq.pxf.plugins.hive.HiveMetadataFetcher</metadata>
+            <outputFormat>BINARY</outputFormat>
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/resources/pxf-profiles-default.xml]: defaultOutputFormat

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveStringPassResolver.java [hunk 6]
```diff
@@ -42,21 +46,36 @@ public HiveStringPassResolver(InputData input) throws Exception {
...
+        if (((ProtocolData) inputData).outputFormat() == OutputFormat.BINARY) {
+            propsString = hiveUserData.getPropertiesString();
+        }
     }
 
     @Override
-    void initSerde(InputData input) {
-        /* nothing to do here */
+    void initSerde(InputData input) throws Exception {
+        if (((ProtocolData) inputData).outputFormat() == OutputFormat.TEXT) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveStringPassResolver.java]: Reverse check.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveAccessor.java (hunk 2 above)
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveAccessor.java]: Yes, even if a table has no partitions, we have "!HNPT!" value.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveLineBreakAccessor.java (hunk 4 above)
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveLineBreakAccessor.java]: partitionKeys always have not-null value.

**Comment by shivzone** [comment]
Please make sure you resolve the conflicts when you merge with master

File: src/bin/gpfusion/gpbridgeapi.c [hunk 7]
```diff
@@ -68,7 +68,6 @@ void	free_token_resources(PxfInputData *inputData);
 Datum gpbridge_import(PG_FUNCTION_ARGS)
 {
 	gpbridge_check_inside_extproto(fcinfo, "gpbridge_import");
-//	ExternalSelectDesc desc = EXTPROTOCOL_GET_SELECTDESC(fcinfo);
```
Comment (kavinderd) [inline src/bin/gpfusion/gpbridgeapi.c]: remove instead of commenting out

File: src/backend/catalog/external/externalmd.c [hunk 8]
```diff
@@ -631,3 +630,79 @@ static int ComputeTypeMod(Oid typeOid, const char *colname, int *typemod, int nT
...
+		appendStringInfo(&formatStr, "formatter 'pxfwritable_import'");
+	}
+	Datum format_opts = DirectFunctionCall1(textin, CStringGetDatum(formatStr.data));
+	pfree(formatStr.data);
+	return format_opts;
+}
+
+/* location - should be an array of text with one element:
+ * pxf://<ip:port/namaservice>/<db>.<path>?Profile=profileName&delimiter=delimiterCode */
+static Datum GetLocationForFormat(char *profile, List *outputFormats, char *pxf_service_address, char *path, char *name, int delimiter)
```
Comment (kavinderd) [inline src/backend/catalog/external/externalmd.c]: Maybe rename path to `db` and name to `path`, the comment and the actual variable names are confusing

Comment (sansanichfb) [inline src/backend/catalog/external/externalmd.c]: Sure

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java [hunk 9]
```diff
@@ -91,12 +94,17 @@ public Field(String name, EnumHawqType type, String sourceType) {
...
         }
 
-        public Field(String name, EnumHawqType type, String sourceType,
-                String[] modifiers) {
+        public Field(String name, EnumHawqType type, String sourceType, String[] modifiers) {
             this(name, type, sourceType);
             this.modifiers = modifiers;
         }
 
+        public Field(String name, EnumHawqType type, boolean isComplexType, String sourceType, String[] modifiers) {
```
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java]: Add isComplexType as the final argument to keep the method overloading structure clean

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/InputData.java [hunk 10]
```diff
@@ -31,6 +31,8 @@
  */
 public class InputData {
 
+    public static final String DELIMITER_KEY = "DELIMITER";
+
```
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/InputData.java]: Remove extra line

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 11]
```diff
@@ -432,32 +447,144 @@ public static String serializePartitionKeys(HiveTablePartition partData) throws
...
+        String[] toks = userData.split(HiveUserData.HIVE_UD_DELIM, EXPECTED_NUM_OF_TOKS);
+
+        if (toks.length != (EXPECTED_NUM_OF_TOKS)) {
+            throw new UserDataException("HiveInputFormatFragmenter expected "
+                    + EXPECTED_NUM_OF_TOKS + " tokens, but got " + toks.length);
+        }
+
+        HiveUserData hiveUserData = new HiveUserData(toks[0], toks[1], toks[2], toks[3], Boolean.valueOf(toks[4]), toks[5]);
+
+            if (supportedSerdes.length > 0) {
```
Comment (kavinderd) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Indentation

**Pull request closed without merging**